### ISSUE TITLE
small gradients with log_softmax when there should be no gradient

### ISSUE THREAD

**Issue by lifengjin** [pytorch/pytorch#16028]
## 🐛 Bug

small gradients with log_softmax when there should be no gradient

## To Reproduce

If you have a perfect uniform vector, and you apply log softmax to it, and backprop the marginal, then the gradients should be 0. For example:
```
a = torch.ones([2, 300],requires_grad=True).float()
p = torch.softmax(a, dim=1).log()
x = p.sum()
x.backward()
a.grad.sum() # = 0
```
But it is not the case if you use log_softmax:

```
a = torch.ones([2, 300],requires_grad=True).float()
p = torch.log_softmax(a, dim=1)
x = p.sum()
x.backward()
a.grad.sum() # = tensor(7.1526e-05)
```
Granted, the gradients are small, but for optimizers like Adam, especially at initialization, no gradient means 0 update, but with a larger learning rate, a small gradient can provide a large update. Does anyone know why there is any gradient at all from log_softmax? The function also behaves correctly when the matrix is small. For example if you do torch.ones([2,100]) etc, then log_softmax is fine.

## Expected behavior

gradients should all be 0

## Environment

Please copy and paste the output from our
[environment collection script](https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py)
(or fill out the checklist below manually).

You can get the script and run it with:
```
wget https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py
# For security purposes, please check the contents of collect_env.py before running it.
python collect_env.py
```
PyTorch version: 1.0.0.dev20190113
Is debug build: No
CUDA used to build PyTorch: 9.0.176

OS: Red Hat Enterprise Linux Server release 7.5 (Maipo)
GCC version: (GCC) 4.8.5
CMake version: version 3.6.3

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: 9.2.88
GPU models and configuration: GPU 0: Tesla V100-PCIE-16GB
Nvidia driver version: 396.37
cuDNN version: Could not collect

Versions of relevant libraries:
[pip] Could not collect
[conda] cuda92                    1.0                           0    pytorch
[conda] magma-cuda80              2.1.0                         5    soumith
[conda] mkl                       2018.0.2                      1    anaconda
[conda] mkl_fft                   1.0.1            py36h3010b51_0    anaconda
[conda] mkl_random                1.0.1            py36h629b387_0    anaconda
[conda] pytorch-nightly           1.0.0.dev20190113 py3.6_cuda9.0.176_cudnn7.4.1_0    pytorch
[conda] torchtext                 0.1.1                     <pip>

**Comment by colesbury** [comment]
It's not a bug. The gradient error is ~1e-7, roughly the [machine epsilon](https://en.wikipedia.org/wiki/Machine_epsilon) of float32 arithmetic. The instability you're describing is an issue with the Adam optimizer.

https://datascience.stackexchange.com/questions/25024/strange-behavior-with-adam-optimizer-when-training-for-too-long
https://openreview.net/pdf?id=ryQu7f-RZ
